tests/end_to_end/end_to_end_lib.py

```diff
@@ -0,0 +1,275 @@
+# debugging.
+# 
+# Then you call start_master_and_chunk_servers function to launch the server
+# cluster:
+#    server_procs = end_to_end_lib.start_master_and_chunk_servers(
+#                       config_filename, log_directory)
+#
+# This function launches the master and chunk servers, and returns the process
+# handle so you can later kill them for cleanup. The default setting is 1 
+# master + 3 chunk servers, and you can configure all fields in the config
+# file via this function (see details for configurable params in the function
+# below). 
+#
+# Once you have finished the above steps, you can start lanching your client
+# binaries and test its end-to-end behavior.  
+#
+# When you test a new end-to-end test or deal with a failure that shows up, 
+# typically you have run:
+#     bazel test -c dbg ...
+# The "-c dbg" is for debugging build and is optional. You may find it hard
+# to locate the generated data and files in bazel as they got generated in
+# sand box (by default). To see the files in sandbox, you have to add 
+# --sandbox_debug option in the above command, and then search under the 
+# bazel build directory (most likely under your ~/.cache). 
+#
+# Alternatively, you could debug your test locally, and to do this, you have
+# to lauch the script in the correct path as the dependencies on files are 
+# expressed in relative path. To do so, first cd into the bazel-bin dir:
+#    cd bazel-bin
+# Then launch your test script in that directory (note that you may have to
+# specify the absolute path of your script, as bazel-bin is a soft-link so
+# "../" does point to a different directory of your cppGFS2.0 root)
+
+master_server_name_prefix = "master_server_"
+chunk_server_name_prefix = "chunk_server_"
+# Used to prevent duplicated port number assignment
+used_port_number = []
+
+# Kill a given process
+def kill_process(proc):
+    if not proc:
+        return
+    if proc.poll() is None:
+        try:
+            os.kill(proc.pid,signal.SIGTERM)
+        except OSError:
+            pass
+
+# Kill all processes in a given list
+def kill_all_processes(procs):
+    for proc in procs:
+        kill_process(proc)
+
+# Pathname to master server's binary
+def master_server_binary():
+    return "src/server/master_server/run_master_server_main"
+
+# Pathname to chunk server's binary
+def chunk_server_binary():
+    return "src/server/chunk_server/run_chunk_server_main"
+
+# Create the test directory for a test case. If the test directory already 
+# exists, a FileExistsError is raised and this either means there is a 
+# duplicated test case name in our test suite, or one has called this function
+# twice in the same test. 
+def setup_test_directory(test_case_name):
+    os.makedirs(test_case_name, exist_ok=True)
+
+# Helper function to generate name for master server for a given server id
+def master_server_name(id):
+    return master_server_name_prefix + "0" + str(id)
```

Comment:
what if it's >9?, so we have master_server_010? why not just master_server_1 to start?

Reply:
This is just an artifact - we opt to have "master_server_01" as a default master name, so I just made this scheme compatible with that. Nothing special here.